src/model/operation/pooling.cc

```diff
@@ -0,0 +1,126 @@
+#include "./pooling.h"
+#include <cmath>
+
+namespace singa {
+
+PoolingHandle::PoolingHandle(const Tensor &input, const std::vector<size_t>& kernel_size,
+                             const std::vector<size_t>& stride, const std::vector<size_t>& padding,
+                             const bool ceil_mode, const std::string pooling_method) {
+  kernel_h = kernel_size[0];
+  kernel_w = kernel_size[1];
+
+  pad_h = padding[0];
+  pad_w = padding[1];
+
+  stride_h = stride[0];
+  stride_w = stride[1];
+
+  batchsize = input.shape(0);
+  channels = input.shape(1);
+  height = input.shape(2);
+  width = input.shape(3);
+
+  pooled_height = 1;
+  if (ceil_mode) {
+    if (stride_h > 0)
+      pooled_height = static_cast<int>(ceil(static_cast<float>(height + 2 * pad_h - kernel_h) / stride_h)) + 1;
+    pooled_width = static_cast<int>(ceil(static_cast<float>(width + 2 * pad_w - kernel_w) / stride_w)) + 1;
+  }
+  else {
+    if (stride_h > 0)
+      pooled_height =
+        static_cast<size_t>((height + 2 * pad_h - kernel_h) / stride_h) + 1;
+    pooled_width =
+      static_cast<size_t>((width + 2 * pad_w - kernel_w) / stride_w) + 1;
+  }
+
+  method = pooling_method;
+  CHECK(method == "MAX" || method == "AVERAGE")
+      << "Padding implemented only for average and max pooling.";
+}
+
+#ifdef USE_CUDNN
+
+CudnnPoolingHandle::CudnnPoolingHandle(const Tensor &input, const std::vector<size_t>& kernel_size,
+                                       const std::vector<size_t>& stride, const std::vector<size_t>& padding,
+                                       const bool ceil_mode, const std::string pooling_method, const bool NaN_prop)
```

Comment:
since you have the pooling_method (for max or avg), you can implement AvgPool2D Layer in the same way.